server/text_generation_server/models/flash_causal_lm.py

```diff
@@ -925,7 +925,11 @@ def __init__(
         assert self.num_kv_heads > 0
 
         if head_size is None:
-            self.head_size = config.hidden_size // config.num_attention_heads
+            if getattr(config, "head_dim", None):
+                # hidden_size / num_attention_heads is wrong in `google/gemma-2-9b-it`
```

Comment:
We can pass the head dim to `FlashCausalLM`, similar to deepseek v2:

https://github.com/huggingface/text-generation-inference/blob/6aebf44f47bc73ac34344fb7b5d941790c11d39d/server/text_generation_server/models/__init__.py#L496

Avoids cluttering this constructor with too much model-specific stuff.

Reply:
Fixed, you are correct (I'm still quite confused about those numbers in the config tho, I think it's because of how they fused qkv.).